cache/blobs.go

```diff
@@ -106,13 +109,45 @@ func computeBlobChain(ctx context.Context, sr *immutableRef, createIfNeeded bool
 			if release != nil {
 				defer release()
 			}
-			desc, err := sr.cm.Differ.Compare(ctx, lower, upper,
-				diff.WithMediaType(mediaType),
-				diff.WithReference(sr.ID()),
-				diff.WithCompressor(compressorFunc),
-			)
-			if err != nil {
-				return nil, err
+			var desc ocispecs.Descriptor
+
+			if forceOvlStr := os.Getenv("BUILDKIT_DEBUG_FORCE_OVERLAY_DIFF"); forceOvlStr != "" {
+				b, err := strconv.ParseBool(forceOvlStr)
+				if err != nil {
+					return nil, errors.Wrapf(err, "invalid boolean in BUILDKIT_DEBUG_FORCE_OVERLAY_DIFF")
+				}
+				if b {
+					computed, ok, err := sr.tryComputeOverlayBlob(ctx, lower, upper, mediaType, sr.ID(), compressorFunc)
+					if !ok {
+						return nil, errors.Errorf("overlay mounts not detected")
+					}
+					if err != nil {
+						return nil, errors.Wrapf(err, "failed to compute overlay diff")
+					}
+					desc = computed
+				} // never call overlay diff if env is false
+			} else if !isTypeWindows(sr) {
+				computed, ok, err := sr.tryComputeOverlayBlob(ctx, lower, upper, mediaType, sr.ID(), compressorFunc)
```

Comment:
the logic looks correct now but if you can, try to call `tryComputeOverlayBlob` only once. Even if it doesn't save lines, because env is(will be) used by the tests it is better if it doesn't go to a different code path. I can imagine a future change that modifies one branch only and we can't catch the error with tests. For the "overlay mounts not detected" case you can print the mount params, both for env and snapshotter case.